rpc/codec/server.go

```diff
@@ -79,6 +90,43 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// Requests need to implement RequestWithUser.
+// We compare the header.User against the client certificate Subject.CommonName.
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
```

Comment:
Rather than performing the authentication here, perhaps serverCodec should contain a `PreHook` that could be invoked on a message before it is dispatched. That would keep the `rpc/codec` package free of dependencies of other cockroach packages.